File: backend/api/endpoints/plugins.py
"""
==========================================================
Easy AI Studio
File    : backend/api/endpoints/plugins.py
Version : 1.0.0
Sprint  : 10
==========================================================
"""

from datetime import datetime
from uuid import uuid4

from fastapi import APIRouter, HTTPException


router = APIRouter()


PLUGINS = []


@router.get("/")
async def get_plugins():

    return {

        "count": len(PLUGINS),

        "plugins": PLUGINS

    }


@router.post("/")
async def install_plugin(data: dict):

    plugin = {

        "id": str(uuid4()),

        "name": data.get("name", "Unknown Plugin"),

        "version": data.get("version", "1.0.0"),

        "author": data.get("author", ""),

        "description": data.get("description", ""),

        "enabled": data.get("enabled", True),

        "installed_at": datetime.now().isoformat()

    }

    PLUGINS.append(plugin)

    return plugin


@router.get("/{plugin_id}")
async def get_plugin(plugin_id: str):

    for plugin in PLUGINS:

        if plugin["id"] == plugin_id:

            return plugin

    raise HTTPException(

        status_code=404,

        detail="Plugin not found."

    )


@router.put("/{plugin_id}")
async def update_plugin(plugin_id: str, data: dict):

    for plugin in PLUGINS:

        if plugin["id"] == plugin_id:

            plugin.update(data)

            return plugin

    raise HTTPException(

        status_code=404,

        detail="Plugin not found."

    )


@router.delete("/{plugin_id}")
async def uninstall_plugin(plugin_id: str):

    for plugin in PLUGINS:

        if plugin["id"] == plugin_id:

            PLUGINS.remove(plugin)

            return {

                "success": True

            }

    raise HTTPException(

        status_code=404,

        detail="Plugin not found."

    )

File: frontend_old/js/plugins.js
/* ==========================================================
   Easy AI Studio
   File: frontend/js/plugins.js
   Version: 1.0.0
   Sprint: 5 - Plugin Manager
   ========================================================== */

"use strict";

/* ==========================================================
   Plugin Manager
   ========================================================== */

const Plugins = {

    storageKey: "plugins",

    registry: [],

    loaded: new Map(),

    /* ======================================================
       Initialize
       ====================================================== */

    init() {

        this.load();

        this.loadEnabledPlugins();

        console.log(

            "Plugin Manager Ready"

        );

    },

    /* ======================================================
       Load Registry
       ====================================================== */

    load() {

        this.registry = Storage.get(

            this.storageKey,

            []

        );

    },

    /* ======================================================
       Save Registry
       ====================================================== */

    save() {

        Storage.set(

            this.storageKey,

            this.registry

        );

    },

    /* ======================================================
       Register Plugin
       ====================================================== */

    register(plugin = {}) {

        const item = {

            id: plugin.id || Utils.uuid(),

            name: plugin.name || "Plugin",

            version: plugin.version || "1.0.0",

            author: plugin.author || "Unknown",

            description: plugin.description || "",

            enabled: plugin.enabled ?? true,

            loaded: false,

            created: Utils.now()

        };

        this.registry.push(item);

        this.save();

        Notification.success(

            "Plugin registered."

        );

        return item;

    },

    /* ======================================================
       Load Enabled Plugins
       ====================================================== */

    loadEnabledPlugins() {

        this.registry.forEach(plugin => {

            if (plugin.enabled) {

                this.loaded.set(

                    plugin.id,

                    plugin

                );

                plugin.loaded = true;

            }

        });

    },

    /* ======================================================
       Enable
       ====================================================== */

    enable(id) {

        const plugin = this.get(id);

        if (!plugin) {

            return;

        }

        plugin.enabled = true;

        plugin.loaded = true;

        this.loaded.set(id, plugin);

        this.save();

    },

    /* ======================================================
       Disable
       ====================================================== */

    disable(id) {

        const plugin = this.get(id);

        if (!plugin) {

            return;

        }

        plugin.enabled = false;

        plugin.loaded = false;

        this.loaded.delete(id);

        this.save();

    },

    /* ======================================================
       Remove
       ====================================================== */

    remove(id) {

        this.registry = this.registry.filter(

            plugin => plugin.id !== id

        );

        this.loaded.delete(id);

        this.save();

    },

    /* ======================================================
       Find
       ====================================================== */

    get(id) {

        return this.registry.find(

            plugin => plugin.id === id

        );

    },

    /* ======================================================
       Get Loaded
       ====================================================== */

    getLoaded() {

        return Array.from(

            this.loaded.values()

        );

    },

    /* ======================================================
       Get All
       ====================================================== */

    all() {

        return this.registry;

    },

    /* ======================================================
       Clear
       ====================================================== */

    clear() {

        this.registry = [];

        this.loaded.clear();

        this.save();

    }

};

/* ==========================================================
   Global
   ========================================================== */

window.Plugins = Plugins;